api.go

```diff
@@ -167,62 +169,180 @@ func (s StorageCallVShardError) Error() string {
 	return fmt.Sprintf("%+v", alias(s))
 }
 
-type StorageResultTypedFunc = func(result interface{}) error
+type StorageResultTypedFunc = func(result ...interface{}) error
 
 type CallOpts struct {
 	VshardMode VshardMode // vshard mode in call
 	PoolMode   pool.Mode
 	Timeout    time.Duration
 }
 
-// revive warns us: time-naming: var CallTimeoutMin is of type time.Duration; don't use unit-specific suffix "Min".
-// But the original lua vshard implementation uses this naming, so we use it too.
-//
-//nolint:revive
-const CallTimeoutMin = time.Second / 2
+// VshardRouterCallMode is a type to represent call mode for Router.Call method.
+type VshardRouterCallMode int
+
+const (
+	// VshardRouterCallModeRO sets a read-only mode for Router.Call.
+	VshardRouterCallModeRO VshardRouterCallMode = iota
+	// VshardRouterCallModeRW sets a read-write mode for Router.Call.
+	VshardRouterCallModeRW
+	// VshardRouterCallModeRE acts like VshardRouterCallModeRO
+	// with preference for a replica rather than a master.
+	// This mode is not supported yet.
+	VshardRouterCallModeRE
+	// VshardRouterCallModeBRO acts like VshardRouterCallModeRO with balancing.
+	VshardRouterCallModeBRO
+	// VshardRouterCallModeBRE acts like VshardRouterCallModeRO with balancing
+	// and preference for a replica rather than a master.
+	VshardRouterCallModeBRE
+)
+
+// VshardRouterCallOptions represents options to Router.Call[XXX] methods.
+type VshardRouterCallOptions struct {
+	Timeout time.Duration
+}
+
+// VshardRouterCallResp represents a response from Router.Call[XXX] methods.
+type VshardRouterCallResp struct {
+	rawMessages []msgpack.RawMessage
+}
+
+// Get returns a response from user defined function as []interface{}.
+func (r VshardRouterCallResp) Get() ([]interface{}, error) {
+	resp := make([]interface{}, len(r.rawMessages))
+	return resp, r.GetTyped(resp)
+}
+
+// GetTyped decodes a response from user defined fuction into custom values.
+func (r VshardRouterCallResp) GetTyped(result []interface{}) error {
+	minLen := len(result)
+	if dataLen := len(r.rawMessages); dataLen < minLen {
+		minLen = dataLen
+	}
+
+	for i := 0; i < minLen; i++ {
+		if err := msgpack.Unmarshal(r.rawMessages[i], &result[i]); err != nil {
+			return fmt.Errorf("failed to decode into result[%d] element #%d of response array: %w", i, i, err)
+		}
+	}
+
+	return nil
+}
 
 // RouterCallImpl Perform shard operation function will restart operation
 // after wrong bucket response until timeout is reached
+// Deprecated: RouterCallImpl is deprecated.
+// See https://github.com/KaymeKaydex/go-vshard-router/issues/110.
+// Use Call, Call[RO, RW, RE, BRO, BRE] methods instead.
```

Comment:
```suggestion
// Use Call, Call with RO, RW, RE, BRO, BRE methods instead.
```
cause call is non generic

Reply:
fixed